[frontend/backend] Authorized members for Feedback (#3181)

lndrtrbn opened the pull request:
### Proposed changes

* [front] Refactoring of form to manage authorized members in workspace to be reusable
* [back] Add new attribute `authorized_members` in Feedback entity
* [back] Add new capability to allow managing authorized members on an entity
* [back] Add migration for the new capability
* [back] Enrich the Feedback API to be able to fetch and edit authorized members
* [front] In Settings > Customization > Feedback, manage the case to edit default authorized members
* [front] In Feedback details, manage new access and add a dialog to manage authorized members
* [back] Update tests
* [back] Refactoring of `attributeDefinition` to not use the value of `mandatoryType` to determine if the attribute can have a default value or not (see [this commit](https://github.com/OpenCTI-Platform/opencti/pull/4918/commits/280897fb491d0f085efd537eec71214da7baab48))

### Related issues

* #3181

### Checklist

<!--
Please submit the source code in a way, where you could honestly say `This code is finished`.
If you feel that there are possibilities for improving the code quality, please do so.
By doing this, you are actively helping us to improve the quality of the entire OpenCTI project.
-->

- [x] I consider the submitted work as finished
- [x] I tested the code for its functionality
- [ ] I wrote test cases for the relevant uses case
- [ ] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

### Further comments

You can find manual testing process in Notion page _Extend autorized_member system to Feedback_ in block _Tests cases_.

**Why refactoring `attributeDefinition` ?**
Until now, we determined if an attribute should be returned to the frontend so we can edit its defaults value using the property `mandatoryType` and it worked well because we didn't have attribute which can be editable and not mandatory. So the condition to return the attribute was "is the attribute mandatory or not ?".
But in our case, authorized members can have a default value and is never mandatory so the condition was broken. Instead of just add a tricky condition based on the name of the attribute we decided to revamp, adding a new property `editDefault` that tell if the attribute can be edited or not.

lndrtrbn commented on opencti-platform/opencti-graphql/src/utils/authorizedMembers.ts:
```diff
@@ -42,23 +43,18 @@ export const getAuthorizedMembers = async (
     return false;
   }
-  if (adminIds.includes(MEMBER_ACCESS_ALL)) { // everyone is admin
+  if (adminIds.includes(MEMBER_ACCESS_ALL) || adminIds.includes(MEMBER_ACCESS_CREATOR)) { // everyone  or creator is admin
```
`CREATOR` is a key used as ID in default values for `authorized_members`. The goal is to define, in default values, a specific role to apply to the user that will create feedbacks. At feedback creation, this `CREATOR` fake ID is replaced by the ID of the real user making the request of creation

lndrtrbn commented:
Rebase ongoing

Kedae commented:
Need to fix the missing creator as admin when getting into manage access
Need to find a way of displaying the EE agreement when clicking on the attributes line in settings

lndrtrbn commented:
@Kedae 
Both points have been done.

@SarahBocognano 
1. Fixed,
2. Fixed,
3. Yep that is not a bug, default values are applied at entity creation (for any attributes) so if you modify the default value it has no impact on already existing entities,
4. We didn't find an easy solution to do this, that's why we explain it inside the info message below the switch input in the authorized members dialog form. But the info message changes depending on if the switch is ON or OFF, maybe that's not a good idea

lndrtrbn commented on opencti-platform/opencti-graphql/src/migrations/1699521799493-capability-authorized-members.js:
```diff
@@ -0,0 +1,16 @@
+import { executionContext, SYSTEM_USER } from '../utils/access';
+import { addCapability } from '../domain/grant';
+
+export const up = async (next) => {
```
Oh, yes, I'll change that when the PR is approved so I'll be sure there is not an other added between 😅

SarahBocognano commented:
Tested again, good for me with the explications of the switch button. I think we can do better in terms of user experience (like maybe only display user/groups/organisations who have manage access in their capabilities in the access restriction management) but for now, the message is clear. 
We can add that information to the documentation for now to avoid confusion/questions.

✅ Good to go

SouadHadjiat commented on opencti-platform/opencti-graphql/src/database/generate-message.js:
```diff
@@ -56,12 +58,26 @@ export const generateDeleteMessage = (instance) => {
     throw UnsupportedError('[OPENCTI] Error generating update message with empty inputs');
   }
+
+  const authorizedMembersIds = patchElements.slice(0, 3).flatMap(([,operations]) => {
```
this whole block looks a bit too complicated for just history logging, do we really need this ?

SouadHadjiat commented on opencti-platform/opencti-graphql/src/database/generate-message.js:
```diff
@@ -56,12 +58,26 @@ export const generateDeleteMessage = (instance) => {
+  if (authorizedMembersIds.length > 0) {
+    members = await internalFindByIds(context, SYSTEM_USER, authorizedMembersIds, {
+      baseData: true,
+      baseFields: ['internal_id', 'name']
```
nice thing to return only the fields we need 👍

I guess patchElements can have mutliple operations with authorized_members ?

lndrtrbn replied:
I'm not sure to understand the questions, you can change/add multiple authorized members at once yes

SouadHadjiat replied:
ah yes ok, that is why I was wondering if we really need to log the authorized members name in history.

SouadHadjiat commented on opencti-platform/opencti-graphql/src/manager/historyManager.ts:
```diff
@@ -117,6 +118,7 @@ const eventsApplyHandler = async (context: AuthContext, events: Array<SseEvent<S
       applicant_id: event.data.origin?.applicant_id,
       timestamp: eventDate,
       context_data: contextData,
+      authorized_members: stix.extensions[STIX_EXT_OCTI].authorized_members,
```
why do we want to save authorized members in history index ?

lndrtrbn replied:
To keep same behavior as for marking and organization sharing systems
